### PR TITLE
Bug 1807154: Persist desktop mode option for tabs when Fenix is reopened

### PR DESCRIPTION
**Pull request by Alexandra-Virvara**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1807154

### CODE REVIEW

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 1]
```diff
@@ -767,10 +767,11 @@ abstract class EngineSession(
      * will be restored.
      *
      * @param state A saved session state.
+     * @param enableDesktopMode True, if desktop mode option is enabled, false otherwise
      * @return true if the engine session has successfully been restored with the provided state,
      * false otherwise.
      */
-    abstract fun restoreState(state: EngineSessionState): Boolean
+    abstract fun restoreState(state: EngineSessionState, enableDesktopMode: Boolean): Boolean
```
Comment (jonalmeida) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt:774]: We shouldn't be passing the flag individually as a param here. What happens when we want to add more things? That shouldn't mean n parameters.

We should get the toggle state from the `EngineSessionState` which each of the engine's implement.

**Review by jonalmeida — CHANGES_REQUESTED** [review]
This looks good! There is one change that I think we should make though. Let me know if my comment isn't clear and I can try to do better.

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 2]
```diff
@@ -336,6 +340,11 @@ class GeckoEngineSession(
             return false
         }
 
+        // TODO fetch the correct value for enableDesktopMode
```
Comment (jonalmeida) [inline android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:343]: Argh, this comment is the place where you've discovered that this solution isn't enough. It looks like we should have GeckoView save and restore the state in their `EngineSessionState` (`GeckoSessionState`) which we can then apply here.

Please file a bug in the GeckoView component to add support for restoring desktop mode. I think your change today can still work with some tweaks.

A possible solution is to set the desktop mode _before_ the `restoreState` is called. This happens in the `CreateEngineSessionMiddleware` - specifically above [this line](https://searchfox.org/mozilla-mobile/rev/9f5ba3d5f8fb052d07447bb3b406d4a3b8ca2d82/firefox-android/android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/CreateEngineSessionMiddleware.kt#119) where we can set the value before instructing the `EngineSession` to load.

That might suffice for now.

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/Engine.kt [hunk 3]
```diff
@@ -114,7 +114,11 @@ interface Engine : WebExtensionRuntime, DataCleanable {
      * @return the newly created [EngineSession].
      */
     @MainThread
-    fun createSession(private: Boolean = false, contextId: String? = null): EngineSession
+    fun createSession(
+        private: Boolean = false,
+        enableDesktopMode: Boolean = false,
```
Comment (jonalmeida) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/Engine.kt:119]: General advice: This is another place to stop and consider if we should be passing this as a separate parameter. We're making a big API change here.

I see from the `GeckoEngineSession` implementation that we're using this value to set the `viewPortMode`, but this isn't the only value needed in order to change into desktop mode (we need to set the `userAgentMode` too). If we were to add the user-agent as another parameter that sounds like we are starting a pattern where we could lead to new params for each new setting in the future.

For the cases of `private` and `contextId` these are explicitly needed internally to create a GeckoSession (basically, they are special). Is enabling desktop mode one of those? I'm not sure.

We might need to consider if this change is necessary or the strategy. See the comment above where I think you've identified the problem already with a possible workaround.

**Comment by Alexandra-Virvara** [comment]
bug for GeckoView: https://bugzilla.mozilla.org/show_bug.cgi?id=1838851

**Comment by gabrielluong** [comment]
:construction: **Pull Requests Cleanup:** :construction: Hi, we're looking to reduce the amount of pull request overhead we currently have. This pull request has been a WIP without any recent activities. Please feel free to reopen this PR at any point when you want to have your changes reviewed. Thank you for your contribution!

**Pull request closed without merging**